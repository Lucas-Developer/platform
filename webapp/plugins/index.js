// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

// EXPERIMENTAL - SUBJECT TO CHANGE

import store from 'stores/redux_store.jsx';

global.window.plugins = {};
global.window.plugins.components = {};

export function registerComponents(components) {
    global.window.plugins.components = {...components, ...global.window.plugins.components};
}

export function initializePlugins() {
    const pluginJson = global.window.mm_config.Plugins || '[]';
    const pluginManifests = JSON.parse(pluginJson);

    pluginManifests.forEach((m) => {
        // Fetch the plugin's bundled js
        const xhrObj = new XMLHttpRequest();
        xhrObj.open('GET', m.bundle_path, false);
        xhrObj.send('');

        // Add the plugin's js to the page
        const script = document.createElement('script');
        script.type = 'text/javascript';
        script.text = xhrObj.responseText;
        document.getElementsByTagName('head')[0].appendChild(script);

        // Initialize the plugin
        console.log('Registering ' + m.id + ' plugin...'); //eslint-disable-line no-console
        const plugin = global.window.plugins[m.id];
        plugin.initialize(registerComponents, store);
        console.log('...done'); //eslint-disable-line no-console
    });
}

