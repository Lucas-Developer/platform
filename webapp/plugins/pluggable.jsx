// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

// EXPERIMENTAL - SUBJECT TO CHANGE

import React from 'react';
import PropTypes from 'prop-types';

export default class Pluggable extends React.PureComponent {
    static propTypes = {
        children: PropTypes.element.isRequired
    }

    render() {
        const child = this.props.children.type;

        if (child == null) {
            return null;
        }

        // Include any props passed to this component or to the child component
        const props = {...this.props, ...this.props.children.props};
        Reflect.deleteProperty(props, 'children');

        // Override the default component with any registered plugin's component
        if (global.window.plugins.components.hasOwnProperty(child.name)) {
            const PluginComponent = global.window.plugins.components[child.name];
            return <PluginComponent {...props}/>;
        }

        return <child {...props}/>;
    }
}
