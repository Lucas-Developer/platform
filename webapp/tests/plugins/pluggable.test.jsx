// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

import React from 'react';
import {mount} from 'enzyme';
import {registerComponents} from 'plugins';

import Pluggable from 'plugins/pluggable.jsx';
import ProfilePopover from 'components/profile_popover.jsx';

class ProfilePopoverPlugin extends React.PureComponent {
    render() {
        return <span>{'ProfilePopoverPlugin'}</span>;
    }
}

describe('plugins/Pluggable', () => {
    test('should match snapshot with overridden component', () => {
        registerComponents({ProfilePopover: ProfilePopoverPlugin});

        const wrapper = mount(
            <Pluggable>
                <ProfilePopover
                    user={{}}
                    src='src'
                />
            </Pluggable>
        );
        expect(wrapper).toMatchSnapshot();

        global.window.plugins.components = {};
    });

    test('should match snapshot with no overridden component', () => {
        const wrapper = mount(
            <Pluggable>
                <ProfilePopover
                    user={{}}
                    src='src'
                />
            </Pluggable>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
