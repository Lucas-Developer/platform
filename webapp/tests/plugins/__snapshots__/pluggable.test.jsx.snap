// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/Pluggable should match snapshot with no overridden component 1`] = `
<Pluggable>
  <child
    hasMention={false}
    isRHS={false}
    src="src"
    user={Object {}}
  />
</Pluggable>
`;

exports[`plugins/Pluggable should match snapshot with overridden component 1`] = `
<Pluggable>
  <ProfilePopoverPlugin
    hasMention={false}
    isRHS={false}
    src="src"
    user={Object {}}
  >
    <span>
      ProfilePopoverPlugin
    </span>
  </ProfilePopoverPlugin>
</Pluggable>
`;
